### ISSUE TITLE
Wrong error on emails in Gitlab-CI

### ISSUE THREAD

**Issue by fauust** [lycheeverse/lychee#136]
Hi,
lychee in GitLab-CI wrongly reports error on email addresses, see:
https://salsa.debian.org/faust/go-team.pages.debian.net/-/jobs/1427125#L150-L151

![screenshot_20210209_154154](https://user-images.githubusercontent.com/2197555/107379625-6806c680-6aed-11eb-84e9-455feeadb237.png)

Here is the yaml part:
```yaml
check:
  stage: check
  script:
    - apt-get update -qq && apt-get install -y -qq wget
    - wget -qO- "https://github.com/lycheeverse/lychee/releases/download/v0.5.0/lychee-v0.5.0-x86_64-unknown-linux-gnu.tar.gz" | tar -xz
    - ./lychee --verbose --exclude="irc://irc.debian.org:6667" --exclude="https://anonscm.debian.org" build/*.html
```

When running it locally, no error is detected:
```console
❯ ./lychee --verbose --exclude="irc://irc.debian.org:6667" --exclude="https://anonscm.debian.org" build/*.html | grep @
✅ git@salsa.debian.org [200 OK]
✅ team+pkg-go@tracker.debian.org [200 OK]
```

Am I missing something?

**Comment by mre** [comment]
So I tested it locally by cloning your repo and running the same command and I can verify that it also works for me. Perhaps the CI server is preventing the check.

* Are the necessary ports open? (Namely SMTP port 25)
* Is an SMTP server configured in your CI pipeline?

Here are some tests that you could try: https://www.sparkpost.com/blog/how-to-check-an-smtp-connection-with-a-manual-telnet-session/
You should be able to run them as a throwaway shell script to get the answers or perhaps you can even log in and run a few tests.

This is lychees entire e-mail checking code:
https://github.com/lycheeverse/lychee/blob/4d5b98978cf929e3072dc09e8e1ab4198865be0a/src/client.rs#L216-L230

So all of the above works but lychee does not, that would probably be a case for https://github.com/reacherhq/check-if-email-exists, which is the library we are using under the hood.

**Comment by fauust** [comment]
Ok, so I guess you pointed to the exact pb, I am pretty sure that the CI server does not allow SMTP output (and it does not surprise me, wouldn't that be a very nice spam vector).
I didn't know that lychee was doing such a deep check for email addresses, I though it would maybe just verify that email syntax looks good and also that there is a MX for the domain. But very nice to have this BTW!

In a CI context I guess that this check will fail most of the time? I did not had time to check if it works on https://github.com/lycheeverse/lychee-action but I will in the next days. Do you have any idea on how to fully disable email check then for CI context?

So I will close this as it's clearly not a bug!

**Comment by mre** [comment]
> Do you have any idea on how to fully disable email check then for CI context?

I've added support for skipping e-mails entirely in #137.
Now you can add a `--exclude-mail` to your lychee call. 

> So I will close this as it's clearly not a bug!

That's fine, though. 😊 Your issue inspired a nice discussion and now we have support for skipping e-mails. 👍

**Comment by fauust** [comment]
wow, that was fast! Thanks a lot!

**Comment by fauust** [comment]
and I have just checked, the email check works fine on GitHub actions:
https://github.com/fauust/docker-systemd/runs/1871010280?check_suite_focus=true

**Comment by fauust** [comment]
Hi @mre, I am trying to implement the latest alpha version in my workflow, it seems that the option to exclude email `--exclude-mail` is not working:
https://salsa.debian.org/faust/go-team.pages.debian.net/-/jobs/1520656

using:
```yaml
...
  - wget -qO- "https://github.com/lycheeverse/lychee/releases/download/v0.5.1-alpha/lychee-v0.5.1-alpha-x86_64-unknown-linux-gnu.tar.gz" | tar -xz
  - ./lychee --verbose --exclude-mail --exclude="irc://irc.debian.org:6667" --exclude="https://anonscm.debian.org" build/**/*.html
```

**Comment by mre** [comment]
The exclusion is working, but the parsing of email addresses happened before exclusion; and the parsing failed. @pawroman fixed that in #177 and the change got merged to `master`. The release of 0.6 is blocked on https://github.com/lycheeverse/lychee/issues/59#issuecomment-798677637 at the moment.

**Comment by mre** [comment]
@fauust, can you give it another shot? Just released `v0.6.0-alpha2` with the fix:
https://github.com/lycheeverse/lychee/releases/tag/v0.6.0-alpha2

This was made possible by https://github.com/lycheeverse/lychee/pull/196 thanks to @lebensterben.

**Comment by fauust** [comment]
Hi!
Nope, still red here :(
https://salsa.debian.org/faust/go-team.pages.debian.net/-/jobs/1542439

**Comment by mre** [comment]
Oh my. 🤦 Forgot to pass the option to the builder. #200 (nice, even number!) should fix this. Added a unit- and and integration test to make sure it never breaks again.

**Comment by fauust** [comment]
I will test again later today!

**Comment by mre** [comment]
Alright! Sorry that you had to hang in there for a while and thanks for the feedback. The remaining error seems to be a true [404 link](https://pkg.go.dev/github.com/Debian/pkg-go-tools/cmd/pgt-gopath), correct? 
We should probably not show the github token message, as it isn't really a github link. It's because our regex is too permissive: https://github.com/lycheeverse/lychee/blob/036145eb989f61b72b1c035e083d423360e356c0/src/client.rs#L266.

It should only match URLs starting with a scheme and `www`:

```
http://github.com
https://github.com
https://www.github.com
www.github.com
github.com
```

and perhaps a few more.
That's unrelated as you said, so I guess we can close this and open a new issue for the remaining work.

**Comment by mre** [comment]
Gave it a shot here: https://github.com/lycheeverse/lychee/pull/201
Let's move the discussion there.

**Comment by fauust** [comment]
BTW, I see a diff between https://salsa.debian.org/faust/go-team.pages.debian.net/-/jobs/1543957 and https://salsa.debian.org/faust/go-team.pages.debian.net/-/jobs/1520656, is this something wanted (verbose is less verbose now)?

**Comment by mre** [comment]
Nope, that wasn't on purpose. Don't know exactly what's going on there. Tried to reproduce it locally with your settings. Everything works as expected (lots of links printed in verbose mode). Here is what I tried:

* ✅ Using v0.6.0-alpha4 on macOS,
* ✅ Compiled from scratch on macOS, everything works as expected (lots of links printed in verbose mode)
* ✅ Using v0.6.0-alpha4 on Debian Linux.

I also tried running it with Docker, and I saw that if I don't set `--tty` (Allocate a pseudo-TTY) as a parameter, I only get the status output. Wild guess: maybe that something changed on the CI system?
I noticed that the two CI runs used a different gitlab runner:
* Running with gitlab-runner 13.10.0 (HEAD)
* Running with gitlab-runner 13.9.0 (HEAD)

Could be that something changed there.
Perhaps you can try to run the job and the standalone binary locally on your machine and compare the results?

**Comment by fauust** [comment]
Your guesses are correct! Salsa gitlab instance was uprated very recently. I will make deeper tests in the next days and keep you informed if I find some something interesting. Closing this as the skip email option seems OK now.
